### PR TITLE
Remove xtensor from geometry code

### PR DESCRIPTION
**Pull request by garth-wells**
Makes operations, and performance implications, clearer.

Also removes xtensor in other places where the benefit is not compelling.

### CODE REVIEW

File: cpp/dolfinx/common/TimeLogger.cpp [hunk 1]
```diff
@@ -7,6 +7,7 @@
 #include "TimeLogger.h"
 #include <dolfinx/common/MPI.h>
 #include <dolfinx/common/log.h>
+#include <iostream>
```
Comment (jorgensd) [inline cpp/dolfinx/common/TimeLogger.cpp:10]: Is this include needed?

File: cpp/dolfinx/geometry/gjk.cpp [hunk 2]
```diff
@@ -6,61 +6,74 @@
...
   {
   case 2:
   {
-    auto s0 = xt::row(s, 0);
-    auto s1 = xt::row(s, 1);
-    const double lm = -xt::sum(s0 * (s1 - s0))() / xt::norm_sq(s1 - s0)();
+    auto s0 = s.subspan(0, 3);
+    auto s1 = s.subspan(3, 3);
+    std::array ds = {s1[0] - s0[0], s1[1] - s0[1], s1[2] - s0[2]};
+    const double lm = -(s0[0] * ds[0] + s0[1] * ds[1] + s0[2] * ds[2])
```
Comment (jorgensd) [inline cpp/dolfinx/geometry/gjk.cpp:32]: Could we add a comment here stating what `lm` is, i.e. dot(s0, ds / |ds|)

File: cpp/dolfinx/geometry/gjk.cpp [hunk 3]
```diff
@@ -6,61 +6,74 @@
...
+    {
+      return std::transform_reduce(
+          x.begin(), x.end(), y.begin(), 0.0, std::plus{},
+          [](auto x, auto y) { return (x - y) * (x - y); });
+    };
+    const double ab2 = length(a, b);
+    const double ac2 = length(a, c);
+    const double bc2 = length(b, c);
+
+    auto helper = [](auto& x, auto& y)
```
Comment (jorgensd) [inline cpp/dolfinx/geometry/gjk.cpp:63]: Maybe add a note saying that this is equivalent to `dot(x, x-y)`?

File: cpp/dolfinx/common/TimeLogger.cpp (hunk 1 above)
Comment (garth-wells) [reply cpp/dolfinx/common/TimeLogger.cpp:10]: Yes, because `std::cout` is used in the file.

File: cpp/dolfinx/geometry/gjk.cpp [hunk 4]
```diff
@@ -69,90 +84,122 @@ nearest_simplex(const xt::xtensor<double, 2>& s)
...
-    xt::xtensor_fixed<double, xt::xshape<3>> vmin = xt::row(s, i);
-    double qmin = xt::norm_sq(vmin)();
+    std::size_t pos = 0;
+    {
+      double norm0 = std::numeric_limits<double>::max();
+      for (std::size_t i = 0; i < s.size(); i += 3)
+      {
+        double norm = 0.0;
+        for (std::size_t k = i; k < i + 3; ++k)
+          norm += s[k] * s[k];
```
Comment (jorgensd) [inline cpp/dolfinx/geometry/gjk.cpp:113]: Is there a reason for not using `transform_reduce` here?

File: cpp/dolfinx/geometry/gjk.cpp [hunk 5]
```diff
@@ -161,89 +208,88 @@ nearest_simplex(const xt::xtensor<double, 2>& s)
...
     throw std::runtime_error("Number of rows defining simplex not supported.");
   }
-  }
 }
 //----------------------------------------------------------------------------
-// Support function, finds point p in bd which maximises p.v
-xt::xtensor_fixed<double, xt::xshape<3>>
-support(const xt::xtensor<double, 2>& bd,
-        const xt::xtensor_fixed<double, xt::xshape<3>>& v)
+// Helper function, finds point p in bd which maximises p.v
```
Comment (jorgensd) [inline cpp/dolfinx/geometry/gjk.cpp:218]: It should be called the "Support function", as this is the name of it in the actual algorithm

File: cpp/dolfinx/geometry/gjk.cpp [hunk 6]
```diff
@@ -161,89 +208,88 @@ nearest_simplex(const xt::xtensor<double, 2>& s)
...
+                              const std::array<double, 3>& v)
 {
   int i = 0;
-  double qmax = xt::sum(xt::row(bd, 0) * v)();
-  for (std::size_t m = 1; m < bd.shape(0); ++m)
+  double qmax = bd[0] * v[0] + bd[1] * v[1] + bd[2] * v[2];
+  for (std::size_t m = 1; m < bd.size() / 3; ++m)
   {
-    double q = xt::sum(xt::row(bd, m) * v)();
+    // q = xt::sum(xt::row(bd, m) * v)();
```
Comment (jorgensd) [inline cpp/dolfinx/geometry/gjk.cpp:226]: Remove leftover comment?

File: cpp/dolfinx/geometry/gjk.cpp (hunk 4 above)
Comment (garth-wells) [reply cpp/dolfinx/geometry/gjk.cpp]: A bit too heavyweight.

**Pull request merged**